discovery/openstack/openstack.go

```diff
@@ -0,0 +1,256 @@
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: conf.IdentityEndpoint,
+		Username:         conf.Username,
+		UserID:           conf.UserID,
+		Password:         conf.Password,
+		TenantName:       conf.TenantName,
+		TenantID:         conf.TenantID,
+		DomainName:       conf.DomainName,
+		DomainID:         conf.DomainID,
+	}
+
+	return &Discovery{
+		authOpts: &opts,
+		region:   conf.Region,
+		interval: time.Duration(conf.RefreshInterval),
+		port:     conf.Port,
+	}, nil
+}
+
+// Run implements the TargetProvider interface.
+func (d *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// Get an initial set right away.
+	tg, err := d.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		select {
+		case ch <- []*config.TargetGroup{tg}:
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	ticker := time.NewTicker(d.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := d.refresh()
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (d *Discovery) refresh() (tg *config.TargetGroup, err error) {
+	t0 := time.Now()
+	defer func() {
+		refreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			refreshFailuresCount.Inc()
+		}
+	}()
+
+	provider, err := openstack.AuthenticatedClient(*d.authOpts)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack session: %s", err)
+	}
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: d.region,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack compute session: %s", err)
+	}
+
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+
+	tg = &config.TargetGroup{
+		Source: fmt.Sprintf("OS_%s", d.region),
+	}
+
+	pagerFIP := floatingips.List(client)
+	floatingIPList := make(map[string][]string)
+
+	err = pagerFIP.EachPage(func(page pagination.Page) (bool, error) {
+		result, err := floatingips.ExtractFloatingIPs(page)
+		if err != nil {
+			log.Warn(err)
+		}
+		for _, ip := range result {
+			// Skip not associated ips
+			if ip.InstanceID != "" {
+				floatingIPList[ip.InstanceID] = append(floatingIPList[ip.InstanceID], ip.IP)
+			}
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("could not describe floating: %s", err)
+	}
+
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, fmt.Errorf("could not extract servers: %s", err)
+		}
+
+		for _, s := range serverList {
+			labels := model.LabelSet{
+				openstackLabelInstanceID: model.LabelValue(s.ID),
+			}
+
+			for _, address := range s.Addresses {
+				md, ok := address.([]interface{})
+				if !ok {
+					log.Warn("Invalid type for addresses, excepted array")
```

Comment:
(sorry, address with double "d" of course, fixed in comment above, but another comment for email notifications)